### ISSUE TITLE
Python/C++ API Parity: tensor autograd API

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#25874]
## 🚀 Feature
We would like to implement the following autograd APIs for C++ tensors, with appropriate tests:
(left: Python API, right: C++ API)

- [x] `torch.Tensor.detach` -> `at::Tensor::detach` (https://github.com/pytorch/pytorch/pull/26251)
- [x] `torch.Tensor.detach_` -> `at::Tensor::detach_` (https://github.com/pytorch/pytorch/pull/26251)
- [x] `torch.Tensor.backward` -> `at::Tensor::backward` (https://github.com/pytorch/pytorch/pull/26150)
- [x] `torch.Tensor.set_data` -> `at::Tensor::set_data` (https://github.com/pytorch/pytorch/pull/26647)
- [x] `torch.Tensor.data` -> `at::Tensor::data` (https://github.com/pytorch/pytorch/pull/26008)
- [x] `torch.Tensor.grad` -> `at::Tensor::grad` (https://github.com/pytorch/pytorch/pull/26150)
- [x] `torch.Tensor.grad_fn` -> `at::Tensor::grad_fn` (https://github.com/pytorch/pytorch/pull/28287)
- [x] `torch.Tensor.output_nr` -> `at::Tensor::output_nr` (https://github.com/pytorch/pytorch/pull/26216)
- [x] `torch.Tensor.is_leaf` -> `at::Tensor::is_leaf` (https://github.com/pytorch/pytorch/pull/26186)
- [x] `torch.Tensor._version` -> `at::Tensor::_version` (https://github.com/pytorch/pytorch/pull/26561)
- [x] `torch.Tensor.register_hook` -> `at::Tensor::register_hook` (https://github.com/pytorch/pytorch/pull/26568, https://github.com/pytorch/pytorch/pull/28287)
  - Ideally, `register_hook` should return a `RemovableHandle` object, which we can call `.remove()` on, similar to the Python version https://github.com/pytorch/pytorch/blob/3b43cfde80d828b39abfd988eb64efcbdb5b61a1/torch/tensor.py#L197-L235
- [x] `torch.Tensor.requires_grad_` -> `at::Tensor::requires_grad_` (https://github.com/pytorch/pytorch/pull/26332)
- [x] `torch.Tensor.retain_grad` -> `at::Tensor::retain_grad` (https://github.com/pytorch/pytorch/pull/33349)
- [x] `torch.Tensor._base` -> `at::Tensor::_base` (https://github.com/pytorch/pytorch/pull/33316)

The current situation is that all these APIs exist in `torch::autograd::Variable` class but not `at::Tensor` class, and the user needs to explicitly cast the C++ tensor to a `torch::autograd::Variable` object in order to use those APIs. What we want to do is to add those APIs in `at::Tensor`, and have the VariableType dispatcher route those API calls to the `torch::autograd::Variable` class internally, so that the user doesn't need to do the explicit casting.

Essentially, we want to following to work (using `data()` as an example):
```cpp
#include <torch/torch.h>

int main()
{
  torch::Tensor t = torch::randn({3, 4});
  torch::Tensor t2 = t.data();
  return 0;
}
```

But currently, only this works:
```cpp
#include <torch/torch.h>
#include <torch/csrc/autograd/variable.h>

int main()
{
  torch::Tensor t = torch::randn({3, 4});
  torch::Tensor t2 = torch::autograd::as_variable_ref(t).variable_data();
  return 0;
}
```

Note that some of the APIs (e.g. `backward`/ `set_data` / `grad`) might already exist in `at::Tensor`, but they might still not have appropriate tests in `test/cpp/api/autograd.cpp`. To make sure things work, we would want to add tests for all of the APIs above.

## Motivation
As part of the Python / C++ API parity work, we would like to provide an identical set of Tensor autograd APIs in C++ as we do in Python. This can address user complaints such as:
- https://github.com/pytorch/pytorch/issues/21899
- https://github.com/yf225/pytorch-cpp-issue-tracker/issues/351
- https://github.com/yf225/pytorch-cpp-issue-tracker/issues/194

## Possible Steps
1. Add entries for the Tensor autograd APIs in `native_functions.yaml` and `Tensor.h`
2. Provide stubs in `aten/src/ATen/native/VariableMethodStubs.cpp` to throw error when the the new API is called on a non-Variable
3. Provide implementation for the new API in `torch/csrc/autograd/VariableTypeManual.cpp`, which should just be calling `as_variable_ref(self).some_api_name(...)` which calls into the Variable methods. (Note that you would be able to find current Python tensor autograd API bindings in `torch/csrc/autograd/python_variable.cpp` (e.g. `THPVariable_get_data`), which may help with creating the right implementation in `VariableTypeManual.cpp`)

`backward()` and `set_data()` are good examples of how the dispatching logic should work, and looking into their implementations is highly recommended.

cc @ezyang @SsnL @albanD @zou3519 @yf225

**Comment by yf225** [comment]
cc. @pbelevich

**Comment by ahmhashesh** [comment]
Can I start by this as my first task?

**Comment by yf225** [comment]
@ahmhashesh Thanks so much for your interest! I think `at::Tensor::_base` can be an easier one to start with, and we can use https://github.com/pytorch/pytorch/pull/26561 as an example for how to implement it.

**Comment by yf225** [comment]
All items in this task have been completed. Closing.

**Comment by yf225** [comment]
cc. @glaringlee
